xwayland/xwm.c

```diff
@@ -80,12 +89,124 @@ static bool xcb_call(struct wlr_xwm *xwm, const char *func, uint32_t line,
 	return false;
 }
 
+static void read_surface_class(struct wlr_xwm *xwm,
+		struct wlr_xwayland_surface *surface, xcb_get_property_reply_t *reply) {
+	if (reply->type != XCB_ATOM_STRING &&
+			reply->type != xwm->atoms[UTF8_STRING]) {
+		return;
+	}
+
+	size_t len = xcb_get_property_value_length(reply);
+	char *class = xcb_get_property_value(reply);
+
+	// Unpack two sequentially stored strings: instance, class
+	size_t instance_len = strnlen(class, len);
+	free(surface->instance);
+	if (len > 0 && instance_len < len) {
+		surface->instance = strndup(class, instance_len);
+		class += instance_len + 1;
+	} else {
+		surface->instance = NULL;
+	}
+	free(surface->class);
+	if (len > 0) {
+		surface->class = strndup(class, len);
+	} else {
+		surface->class = NULL;
+	}
+
+	wlr_log(L_DEBUG, "XCB_ATOM_WM_CLASS: %s %s", surface->instance, surface->class);
+	wl_signal_emit(&surface->events.set_class, surface);
+}
+
+static void read_surface_title(struct wlr_xwm *xwm,
+		struct wlr_xwayland_surface *surface, xcb_get_property_reply_t *reply) {
+	if (reply->type != XCB_ATOM_STRING &&
+			reply->type != xwm->atoms[UTF8_STRING]) {
+		return;
+	}
+
+	// TODO: if reply->type == XCB_ATOM_STRING, uses latin1 encoding
+	// if reply->type == xwm->atoms[UTF8_STRING], uses utf8 encoding
+
+	size_t len = xcb_get_property_value_length(reply);
+	char *title = xcb_get_property_value(reply);
+
+	free(surface->title);
+	if (len > 0) {
+		surface->title = strndup(title, len);
+	} else {
+		surface->title = NULL;
+	}
+
+	wlr_log(L_DEBUG, "XCB_ATOM_WM_NAME: %s", surface->title);
+	wl_signal_emit(&surface->events.set_title, surface);
+}
+
+static void read_surface_parent(struct wlr_xwm *xwm,
+		struct wlr_xwayland_surface *surface, xcb_get_property_reply_t *reply) {
+	if (reply->type != XCB_ATOM_WINDOW) {
+		return;
+	}
+
+	xcb_window_t *xid = xcb_get_property_value(reply);
+	if (xid != NULL) {
+		surface->parent = lookup_surface_any(xwm, *xid);
+	} else {
+		surface->parent = NULL;
+	}
+
+	wlr_log(L_DEBUG, "XCB_ATOM_WM_TRANSIENT_FOR: %p", xid);
+	wl_signal_emit(&surface->events.set_parent, surface);
+}
```

Comment:
What is the WM supposed to do with the parent? Anything?

Reply:
I know some compositors move the parent to foreground (but under the child) when the child is activated. Some might also add a semi-transparent overlay to the parent while it has some children. I added the signal for completeness, but that's maybe not needed (at least for now).